codegen/shim/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+#![doc = include_str!("../README.md")]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![cfg_attr(
+    feature = "doc",
+    deny(rustdoc::broken_intra_doc_links, rustdoc::private_intra_doc_links)
+)]
+#![deny(
+    nonstandard_style,
+    rust_2018_idioms,
+    trivial_casts,
+    trivial_numeric_casts
+)]
+#![forbid(non_ascii_idents, unsafe_code)]
+#![warn(
+    deprecated_in_future,
+    missing_copy_implementations,
+    missing_debug_implementations,
+    missing_docs,
+    unreachable_pub,
+    unused_import_braces,
+    unused_labels,
+    unused_qualifications,
+    unused_results
+)]
+
+use arcana_codegen_impl as codegen;
+use proc_macro::TokenStream;
+
+/// Macro for deriving [`Event`] on enums.
+///
+/// For structs consider using [`#[derive(Versioned)]`](macro@VersionedEvent).
+///
+/// This macro ensures that every combination of [`Event::name`][0] and
+/// [`Event::version`][1] are unique. The only limitation is that all the
+/// underlying [`Event`] or [`Versioned`] impls should be derived too.
+///
+/// > __WARNING:__ Currently may not work with complex generics using where
+/// >              clause because of `const` evaluation limitations. Should be
+/// >              lifted once [rust-lang/rust#57775] is resolved.
+///
+/// # Variant attributes
+///
+/// #### `#[event(ignore)]` (optional)
+///
+/// Aliases: `#[event(skip)]`
+///
+/// Use this on a particular enum variant to completely ignore it in code
+/// generation.
+///
+/// > __WARNING:__ Calling [`Event::name()`][0] or [`Event::version()`][1] on
+/// >              ignored variants will result in [`unreachable!`] panic.
+///
+/// # Example
+///
+/// ```rust,compile_fail,E0080
+/// # use arcana::es::{event, Event};
+/// #
+/// #[derive(event::Versioned)]
+/// #[event(name = "chat", version = 1)]
+/// struct ChatEvent;
+///
+/// #[derive(event::Versioned)]
+/// #[event(name = "file", version = 1)]
+/// struct FileEvent;
+///
+/// #[derive(Event)]
+/// enum AnyEvent {
+///     Chat(ChatEvent),
+///     File(FileEvent),
+/// }
+///
+/// // This fails to compile as contains `FileEvent` duplicated.
+/// #[derive(Event)]
+/// enum DuplicatedEvent {
+///     Any(AnyEvent),
+///     File(FileEvent),
```

Comment:
@ilslv it seems that exactly this situation should be legit.

I guess it should be allowed to use the same event type multiple times. This allows us easily to merge sets of events.

What should be forbidden really - is using the same `event_name` + `event_ver` pair in __different__ Rust types.

Let's merge this PR as it is, and resolve the described issue in a separate one.